File: framework/languages/java/samples/java/Bingo/Client/src/main/java/systems/zlink/samples/bingo/client/BingoClientApp.java
package systems.zlink.samples.bingo.client;

import java.util.ArrayList;
import java.util.List;
import systems.zlink.samples.bingo.shared.configuration.SampleTimings;
import systems.zlink.samples.bingo.shared.contracts.Messages;

public final class BingoClientApp {
    private final BingoClientOptions options;

    public BingoClientApp(BingoClientOptions options) {
        this.options = options;
    }

    public void run() throws Exception {
        List<BingoPlayerClient> clients = new ArrayList<>();
        for (int i = 1; i <= options.playerCount(); i++) {
            BingoPlayerClient client = new BingoPlayerClient("player-" + i);
            await("connect " + client.playerId(), client.connectAsync());
            Messages.AuthenticateRes auth =
                await("authenticate " + client.playerId(), client.authenticateAsync());
            require(auth.actorId().equals(client.playerId()), "session authenticate reply mismatch");
            clients.add(client);
        }

        Messages.MatchBingoRes firstMatch =
            await("match " + clients.getFirst().playerId(), clients.getFirst().matchAsync());
        require(firstMatch.state().status().equals("WaitingForPlayers"),
            "first match must wait for another player");

        List<Messages.MatchBingoRes> matches = new ArrayList<>();
        matches.add(firstMatch);
        for (int i = 1; i < clients.size(); i++) {
            matches.add(await("match " + clients.get(i).playerId(), clients.get(i).matchAsync()));
        }
        require(matches.stream().map(Messages.MatchBingoRes::roomId).distinct().count() == 1,
            "all match requests must return the same room");
        require(matches.getFirst().state().hostActorId().equals(clients.getFirst().playerId()),
            "first joined actor must become host");
        require(matches.getLast().state().status().equals("Running"),
            "second match must start the room automatically");

        waitForStarted(clients);
        Messages.SubmitBingoCardRes secondCard = await(
            "submit card " + clients.get(1).playerId(),
            clients.get(1).submitCardAsync(firstMatch.roomId(), List.of(1, 2, 3, 4, 0, 5, 6, 7, 8)));
        require(secondCard.state().players().get(1).card().size() == 9,
            "second player card must be stored immediately");
        Messages.SubmitBingoCardRes firstCard = await(
            "submit card " + clients.getFirst().playerId(),
            clients.getFirst().submitCardAsync(firstMatch.roomId(), List.of(1, 9, 10, 11, 0, 12, 13, 14, 15)));
        require(firstCard.state().players().getFirst().card().size() == 9,
            "first player card must be stored immediately");

        Messages.BingoRoomState ended = waitForEnded(clients);
        require(ended.status().equals("Finished"), "room must finish through timer draws");
        require(!ended.winners().isEmpty(), "deterministic sample must finish with at least one winner");
        require(ended.players().stream().allMatch(player -> player.card().size() == 9),
            "each player card must contain 9 cells");
        require(ended.players().stream().allMatch(player -> Boolean.TRUE.equals(player.marks().get(4))),
            "center free cell must start marked");
        require(clients.stream().allMatch(client -> !client.inbox().started().isEmpty()),
            "each connector must receive game-start push");
        require(clients.stream().allMatch(client -> !client.inbox().drawn().isEmpty()),
            "each connector must receive draw push");
        require(clients.stream().allMatch(client -> !client.inbox().ended().isEmpty()),
            "each connector must receive game-ended push");
    }

    private static void waitForStarted(List<BingoPlayerClient> clients) throws Exception {
        long deadline = System.nanoTime() + SampleTimings.RequestTimeout.toNanos();
        while (System.nanoTime() < deadline) {
            for (BingoPlayerClient client : clients) {
                await("dispatch " + client.playerId(), client.dispatchAsync());
            }
            if (clients.stream().allMatch(client -> !client.inbox().started().isEmpty())) {
                return;
            }
            Thread.onSpinWait();
        }
        throw new IllegalStateException("Timed out waiting for BingoGameStartedNotify");
    }

    private static Messages.BingoRoomState waitForEnded(List<BingoPlayerClient> clients) throws Exception {
        long deadline = System.nanoTime() + SampleTimings.RequestTimeout.toNanos();
        while (System.nanoTime() < deadline) {
            for (BingoPlayerClient client : clients) {
                await("dispatch " + client.playerId(), client.dispatchAsync());
            }
            if (clients.stream().allMatch(client -> !client.inbox().ended().isEmpty())) {
                return clients.stream()
                    .flatMap(client -> client.inbox().ended().stream())
                    .reduce((first, second) -> second)
                    .orElseThrow()
                    .state();
            }
            Thread.onSpinWait();
        }
        throw new IllegalStateException("Timed out waiting for BingoGameEndedNotify");
    }

    private static <T> T await(String action, java.util.concurrent.CompletionStage<T> stage)
        throws Exception {
        try {
            return SampleAsync.await(stage);
        } catch (Exception ex) {
            throw new IllegalStateException("Bingo sample failed during " + action, ex);
        }
    }

    private static void require(boolean condition, String message) {
        if (!condition) {
            throw new IllegalStateException(message);
        }
    }

}

File: framework/languages/java/samples/kotlin/Bingo/Client/src/main/kotlin/systems/zlink/samples/kotlin/bingo/client/BingoClientApp.kt
package systems.zlink.samples.kotlin.bingo.client

import java.util.concurrent.CompletionException
import kotlinx.coroutines.delay
import systems.zlink.samples.kotlin.bingo.shared.configuration.SampleTimings
import systems.zlink.samples.kotlin.bingo.shared.contracts.BingoRoomState

class BingoClientApp(
    private val options: BingoClientOptions,
) {
    suspend fun run(drawSequence: List<Int>) {
        val clients = mutableListOf<BingoPlayerClient>()
        for (index in 1..options.playerCount) {
            val client = BingoPlayerClient("player-$index")
            runAction("connect ${client.playerId}") { client.connect() }
            val auth = runAction("authenticate ${client.playerId}") { client.authenticate() }
            require(auth.actorId == client.playerId) { "session authenticate reply mismatch" }
            clients += client
        }

        val firstMatch = runAction("match ${clients.first().playerId}") {
            clients.first().match()
        }
        requireRejected("host start must be rejected before four players join") {
            runAction("premature start ${clients.first().playerId}") {
                clients.first().start(firstMatch.roomId)
            }
        }

        val matches = mutableListOf(firstMatch)
        for (client in clients.drop(1)) {
            matches += runAction("match ${client.playerId}") { client.match() }
        }
        require(matches.map { it.roomId }.distinct().size == 1) {
            "all match requests must return the same room"
        }
        require(matches.first().state.hostActorId == clients.first().playerId) {
            "first joined actor must become host"
        }

        requireRejected("non-host start must be rejected") {
            runAction("non-host start ${clients[1].playerId}") { clients[1].start(firstMatch.roomId) }
        }
        val started = runAction("host start ${clients.first().playerId}") {
            clients.first().start(firstMatch.roomId)
        }
        require(started.state.status == "Running") {
            "host start must put room into Running status"
        }

        val ended = waitForEnded(clients)
        require(ended.status == "Finished") { "room must finish through timer draws" }
        require(ended.winners.size > 1) {
            "deterministic sample must include same-sequence winners"
        }
        require(ended.players.all { it.card.size == 25 }) {
            "each player card must contain 25 cells"
        }
        require(ended.players.all { it.marks[12] }) {
            "center free cell must start marked"
        }
        require(clients.all { it.inbox.started().isNotEmpty() }) {
            "each connector must receive game-start push"
        }
        require(clients.all { it.inbox.drawn().isNotEmpty() }) {
            "each connector must receive draw push"
        }
        require(clients.all { it.inbox.ended().isNotEmpty() }) {
            "each connector must receive game-ended push"
        }
        clients.forEach(BingoPlayerClient::close)
    }

    private suspend fun waitForEnded(clients: List<BingoPlayerClient>): BingoRoomState {
        val deadline = System.nanoTime() + SampleTimings.RequestTimeout.toNanos()
        while (System.nanoTime() < deadline) {
            clients.forEach { it.dispatch() }
            if (clients.all { it.inbox.ended().isNotEmpty() }) {
                return clients
                    .flatMap { it.inbox.ended() }
                    .last()
                    .state
            }
            delay(10)
        }
        throw IllegalStateException("Timed out waiting for BingoGameEndedNotify")
    }

    private suspend fun requireRejected(message: String, action: suspend () -> Unit) {
        try {
            action()
        } catch (ex: CompletionException) {
            return
        } catch (ex: IllegalStateException) {
            return
        }
        throw IllegalStateException(message)
    }

    private suspend fun <T> runAction(action: String, block: suspend () -> T): T {
        try {
            return block()
        } catch (ex: Exception) {
            throw IllegalStateException("Bingo sample failed during $action", ex)
        }
    }
}
